### PR TITLE
C and ScaLAPACK-like API for generalised eigensolver

### LINKED ISSUE #947: Add C/ScaLAPACK API for generalized eigensolver
Follow-up to #886.

### PR DESCRIPTION
**Pull request by RMeli**
Close #947.

### CODE REVIEW

File: test/unit/eigensolver/test_gen_eigensolver.cpp [hunk 1]
```diff
@@ -66,76 +67,14 @@ const std::vector<std::tuple<SizeType, SizeType, SizeType>> sizes = {
...
-
 template <class T, Backend B, Device D, Allocation allocation, class... GridIfDistributed>
 void testGenEigensolver(const blas::Uplo uplo, const SizeType m, const SizeType mb,
                         GridIfDistributed... grid) {
   constexpr bool isDistributed = (sizeof...(grid) == 1);
 
   const TileElementSize block_size(mb, mb);
 
-  auto create_reference = [&]() -> auto{
+  auto create_reference = [&]() -> auto {
```
Comment (RMeli) [inline test/unit/eigensolver/test_gen_eigensolver.cpp:77]:
```suggestion
  auto create_reference = [&]() -> auto{
```

File: test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+  auto dlaf_context = c_api_test_inititialize<api>(grid);
+
+  // In normal use the runtime is resumed by the C API call
+  // The pika runtime is suspended by dlaf_initialize
+  // Here we need to resume it manually to build the matrices with DLA-Future
+  pika::resume();
+
+  const TileElementSize block_size(mb, mb);
+
+  auto create_reference = [&]() -> auto {
```
Comment (RMeli) [inline test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp:79]:
```suggestion
  auto create_reference = [&]() -> auto{
```

**Review by RMeli — COMMENTED** [review]
Formatting.

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

File: test/unit/eigensolver/test_gen_eigensolver.cpp [hunk 3]
```diff
@@ -66,68 +67,6 @@ const std::vector<std::tuple<SizeType, SizeType, SizeType>> sizes = {
     {34, 8, 3},  {32, 6, 3}                                   // m > mb, sub-band
 };
 
-template <class T, Device D, class... GridIfDistributed>
-void testGenEigensolverCorrectness(const blas::Uplo uplo, Matrix<const T, Device::CPU>& reference_a,
-                                   Matrix<const T, Device::CPU>& reference_b,
-                                   EigensolverResult<T, D>& ret, GridIfDistributed... grid) {
```
Comment (RMeli) [inline test/unit/eigensolver/test_gen_eigensolver.cpp:72]: Moved to `include/dlaf_test/test_gen_eigensolver_correctness.h`.

File: test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+void testGenEigensolver(const blas::Uplo uplo, const SizeType m, const SizeType mb,
+                        CommunicatorGrid grid, int dlaf_context) {
+  // In normal use the runtime is resumed by the C API call
+  // The pika runtime is suspended by dlaf_initialize
+  // Here we need to resume it manually to build the matrices with DLA-Future
+  pika::resume();
+
+  const TileElementSize block_size(mb, mb);
+
+  auto create_reference = [&]() -> auto{
```
Comment (msimberg) [inline test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp:74]: Very minor, is the `auto` needed?
```suggestion
  auto create_reference = [&]() {
```

**Review by msimberg — APPROVED** [review]
Looks good to me!

**Review by rasolca — APPROVED** [review]

File: test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp (hunk 4 above)
Comment (RMeli) [reply test/unit/c_api/eigensolver/test_gen_eigensolver_c_api.cpp]: I don't think so. It was copy-pasted from `test_gen_eigensolver.cpp`, so I removed it there as well.

**Comment by RMeli** [comment]
cscs-ci run

**Comment by RMeli** [comment]
cscs-ci run

**Review by msimberg — APPROVED** [review]

**Pull request merged**